libs/constrained-generators/src/Constrained/GenT.hs

```diff
@@ -62,82 +118,115 @@ fatalError1 s = fatalError (pure s)
-  deriving (Ord, Eq, Show, Functor)
-
-instance Applicative GE where
-  pure = Result []
-  (<*>) = ap
-
-instance Monad GE where
-  FatalError es err >>= _ = FatalError es err
-  GenError es err >>= _ = GenError es err
-  Result _ a >>= k = k a
+-- GE instance
 
 instance MonadGenError GE where
-  genError neStr = GenError [] neStr
-  fatalError neStr = FatalError [] neStr
-  explain nes ge = case ge of
-    GenError es' err -> GenError (nes : es') err
-    FatalError es' err -> FatalError (nes : es') err
-    Result es' a -> Result (nes : es') a
+  genError msg = GenError (pure msg)
+  genErrors msgs = GenError msgs
+  fatalError msg = FatalError (pure msg)
+  fatalErrors msgs = FatalError msgs
+  explain m (GenError ms) = GenError (m <| ms)
+  explain m (FatalError ms) = FatalError (m <| ms)
+  explain _ (Result x) = Result x
+
+-- GenT instance
 
+-- | calls to genError and fatalError, add the stacked messages in the monad.
 instance MonadGenError m => MonadGenError (GenT m) where
-  genError es = GenT $ \_ -> pure $ genError es
-  fatalError es = GenT $ \_ -> pure $ fatalError es
-  explain es gen = GenT $ \mode -> fmap (explain es) (runGenT gen mode)
+  genError e = GenT $ \(_, xs) -> pure $ genErrors (add e xs)
+  genErrors es = GenT $ \(_, xs) -> pure $ genErrors (cat es xs)
 
-instance MonadGenError m => MonadFail (GenT m) where
-  fail s = genError (pure s)
+  -- Perhaps we want to turn genError into fatalError, if mode_ is Strict?
+  fatalError e = GenT $ \(_, xs) -> pure $ fatalErrors (add e xs)
+  fatalErrors es = GenT $ \(_, xs) -> pure $ fatalErrors (cat es xs)
 
-catGEs :: MonadGenError m => [GE a] -> m [a]
-catGEs [] = pure []
-catGEs (Result _ a : ges) = (a :) <$> catGEs ges
-catGEs (GenError {} : ges) = catGEs ges
-catGEs (FatalError es e : _) =
-  runGE $ FatalError es e
-
-fromGE :: (NE.NonEmpty String -> a) -> GE a -> a
-fromGE _ (Result _ a) = a
-fromGE a (GenError [] e) = a e
-fromGE a (GenError es e) = a $ foldr1 (<>) es <> e
-fromGE _ (FatalError es e) =
-  error . ("\n" ++) . unlines $ concat (map NE.toList es) ++ (NE.toList e)
+  -- Perhaps we want to turn fatalError into genError, if mode_ is Loose?
+  explain e (GenT f) = GenT $ \(mode, es) -> fmap (explain e) (f (mode, es))
 
-errorGE :: GE a -> a
-errorGE mge = case mge of
-  FatalError xs x -> error $ mkErrorMsg xs x
-  GenError xs x -> error $ mkErrorMsg xs x
-  Result _ x -> x
+-- Perhaps we want to turn fatalError into genError, if mode_ is Loose?
+
+-- ====================================================
+-- useful operations on NonEmpty
+
+add :: NonEmpty a -> [NonEmpty a] -> NonEmpty (NonEmpty a)
+add a [] = pure a
+add a (x : xs) = a <| (x :| xs)
+
+cat :: NonEmpty (NonEmpty a) -> [NonEmpty a] -> NonEmpty (NonEmpty a)
+cat a [] = a
+cat a (x : xs) = a <> (x :| xs)
+
+-- | Similar to (concat :: [[x]] -> [x]) for lists
+concatNE :: NonEmpty (NonEmpty t) -> NonEmpty t
+concatNE (x :| []) = x
+concatNE (x :| (y : ys)) = x <> concatNE (y :| ys)
```

Comment:
Same as previous PR can be replaced by `sconcat`
```suggestion
```

Reply:
done